game-core/src/main/java/games/strategy/triplea/ai/tree/StepUnits.java

```
@@ -0,0 +1,446 @@
+        }
+        for (int i = 0; i < entry.getValue() - 1; i++) {
+          newTargets.add(0, entry.getKey());
+        }
+      }
+    }
+    return newTargets;
+  }
+
+  boolean noMoreFriendlies() {
+    return countOfFriendlyDamagedOrDead() == friendlyUnits.size() && retreatedFriendly.isEmpty();
+  }
+
+  boolean noMoreEnemies() {
+    return countOfEnemyDamagedOrDead() == enemyUnits.size() && retreatedEnemy.isEmpty();
+  }
+
+  int countOfFriendliesNotDamagedOrDead() {
+    return friendlyUnits.size() - countOfFriendlyDamagedOrDead();
+  }
+
+  int countOfFriendliesNotDead() {
+    return friendlyUnits.size() - countOfFriendlyDead();
+  }
+
+  int countOfEnemiesNotDamagedOrDead() {
+    return enemyUnits.size() - countOfEnemyDamagedOrDead();
+  }
+
+  int countOfEnemiesNotDead() {
+    return enemyUnits.size() - countOfEnemyDead();
+  }
+
+  int countOfFriendlyDamagedOrDead() {
+    return friendlyWaitingToDieBits.cardinality();
+  }
+
+  int countOfFriendlyDead() {
+    return friendlyBits.cardinality();
+  }
+
+  int countOfEnemyDamagedOrDead() {
+    return enemyWaitingToDieBits.cardinality();
+  }
+
+  int countOfEnemyDead() {
+    return enemyBits.cardinality();
+  }
+
+  int countOfFriendlyHitPoints() {
+    return friendlyHitPoints.values().stream().reduce(0, Integer::sum);
+  }
+
+  /** Store whether the units were alive */
+  void recordChances() {
+    for (final Unit unit : getAliveFriendly()) {
+      friendlyUnitsChances.put(unit, 1.0);
+    }
+    for (final Unit unit : retreatedFriendly) {
+      friendlyUnitsChances.put(unit, 1.0);
+    }
+    for (final Unit unit : getAliveEnemy()) {
+      enemyUnitsChances.put(unit, 1.0);
+    }
+    for (final Unit unit : retreatedEnemy) {
+      enemyUnitsChances.put(unit, 1.0);
+    }
+  }
+
+  void updateUnitChances(final StepUnits child, final double probability) {
+    for (final Map.Entry<Unit, Double> entry : child.enemyUnitsChances.entrySet()) {
+      friendlyUnitsChances.compute(
```

This case should never really occur if you ask me, Double should never be null, so I'd rather throw an exception here

---

It can occur and does fairly often.  It happens when two child branches have different units hit.   This is merging the branch results so one branch will have a unit that has a probability for a unit and the other won't.  enemyUnitsChances/friendlyUnitsChances only has entries if there is a chance the unit might survive.

---

Ah, nevermind.
I thought you were updating the same map you were reading from, which isn't the case 👍